rfcs/59-from_reflect_ergonomics.md

```diff
@@ -0,0 +1,284 @@
+
+> As an added benefit, we only need to parse the `ReflectDeriveData` once, as opposed to twice for both reflect derives.
+
+### Implementing: Dynamic Conversions
+
+Implementing **[Dynamic Conversions](#dynamic-conversions)** mainly involves the creation of a `ReflectFromReflect` type. This was previously explored in https://github.com/bevyengine/bevy/pull/4147, however, it was abandoned because of the issues this RFC aim to fix.
+
+This type data should automatically be registered if `FromReflect` is automatically derived.
+
+### Implementing: Deserialization Clarity
+
+**[Deserialization Clarity](#deserialization-clarity)** can readily be implemented by adding a private field to `ReflectDeserializer`:
+
+```rust
+pub struct ReflectDeserializer<'a> {
+  registry: &'a TypeRegistry,
+  auto_convert: bool // <- New
+}
+```
+
+This should default to `true` and be changeable using a new method:
+
+```rust
+impl<'a> ReflectDeserializer<'a> {
+  pub fn set_auto_convert(&mut self, value: bool) {
+    self.auto_convert = value;
+  }
+}
+```
+
+Once the Dynamic is deserialized by the visitor, it will check this value to see whether it needs to perform the conversion. This is done using the same method as shown in **[Dynamic Conversions](#dynamic-conversions)**. However, rather than unwrapping, we will return a `Result` that gives better error messages that point to `FromReflect` and/or `ReflectFromReflect`.
+
+## Drawbacks
+
+* Unidiomatic. One of the biggest drawbacks of this change is that it is not very standard to have opt-out mechanisms for derives. They mainly act in an additive fashion. Thus, it could be confusing.
+
+* Increased code generation. By automatically deriving `FromReflect` we increase the amount of code that needs to be generated per type— namely for reflectable types that *really* don't need `FromReflect` (such as for a single binary that doesn't need serialization and is not meant to be a library).
+
+  <details>
+  <summary>Rebuttal</summary>  
+
+    In most cases, however, this is code that *should* be generated as explained in **[Derive Complexity](#derive-complexity)**. Additionally, the code generation is rather small with each field in a struct only generating around 1–4 lines of code.
+
+  </details>
+
+## Rationale and alternatives
+
+The biggest alternative is doing nothing. Again, most of these are ergonomics and ecosystem wins. The engine will work pretty much the same with or without these changes. The hope is that we can make reflection less intimidating, confusing, and frustrating, while making it simpler, more intuitive, and more consistent across crates.
+
+One concern about integrating the `FromReflect` derive into the `Reflect` derive is that we might want to support something like a `FromWorldReflect`. However, using the opt-out method this should still be possible.
+
+## Unresolved questions
+
+- Should `ReflectFromReflect` be replaced by a method on `TypeRegistration`, such as `TypeRegistration::construct`?
```

Comment:
I don't think so, but I wouldn't be opposed to adding `construct()` as a helper function which wraps over `ReflectFromReflect`.

Reply:
Yeah I agree that we should keep RFR. I decided to leave out the `Future Possibilities` section, but in it I originally described adding a `ReflectFromWorldReflect` type data to support a proposed `FromWorldReflect` trait. And my thought was that this would pair nicely with RFR without needing to create another method on `TypeRegistration`. (Left out that section because it seemed a bit off-topic, though, haha.)

And also, by keeping RFR we keep open the possibility of cloning and storing the type data outside the registry if needed.